cherab/tools/inversions/opencl/opencl_utils.py

```diff
@@ -0,0 +1,173 @@
+            if verbose:
+                print("Selected OpenCL device: %s %s\n" % (device.get_info(cl.device_info.VENDOR), device.get_info(cl.device_info.NAME)))
+            return device
+    print("\nThere are no devices of specified type\n")
+
+    return None
+
+
+def device_select(platfrom_id=None, device_id=None, device_type=cl.device_type.GPU | cl.device_type.ACCELERATOR, verbose=True):
+    """
+    OpenCL device selector. Returns the most powerfull OpenCL device availabe if device_type is GPU or accelerator
+    or the first OpenCL device available if device_type is CPU.
+
+    :param int platfrom_id: OpenCL platform ID, defaults to `platfrom_id=None`.
+    :param int device_id: OpenCL device ID (in the selected OpenCL platform), defaults to `device_id=None`.
+    :param pyopencl.device_type device_type: OpenCL device type (GPU, ACCELERATOR, etc.).
+        Default value: `device_type=pyopencl.device_type.GPU | pyopencl.device_type.ACCELERATOR`.
+    :param bool verbose: Verbose output, defaults to `verbose=True`.
+
+    :return: The pyopencl.Device instance corresponding to the selected OpenCL device.
+    """
+    platforms = cl.get_platforms()
+    n_platforms = len(platforms)
+    if device_type == cl.device_type.DEFAULT:
+        device_type = cl.device_type.GPU | cl.device_type.ACCELERATOR
+    non_gpu_device = not (cl.device_type.GPU | cl.device_type.ACCELERATOR) & device_type
+    if platfrom_id is None:
+        if non_gpu_device:
+            return get_first_device(platforms, device_type, verbose)
+        return get_best_gpu(platforms, device_type, verbose)
+    if platfrom_id < n_platforms:
+        platform = platforms[platfrom_id]
+        devices = platform.get_devices(device_type=device_type)
+        n_devices = len(devices)
+        if device_id is None:
+            if non_gpu_device:
+                return get_first_device([platform], device_type, verbose)
+            return get_best_gpu([platform], device_type, verbose)
+        if device_id < n_devices:
+            device = devices[device_id]
+            if verbose:
+                print("Selected OpenCL device: %s %s" % (device.get_info(cl.device_info.VENDOR), device.get_info(cl.device_info.NAME)))
+            if non_gpu_device:
+                return (device, 0)
+            return (device, get_flops(device, verbose))
+        warnings.warn('%s platform has %d devices of specified type\n' % (platform.get_info(cl.platform_info.NAME), n_platforms))
+        if non_gpu_device:
+            return get_first_device([platform], device_type, verbose)
+        return get_best_gpu([platform], device_type, verbose)
+    warnings.warn('System has only %d OpenCL platforms\n' % n_platforms)
+    if non_gpu_device:
+        return get_first_device(platforms, device_type, verbose)
+
+    return get_best_gpu(platforms, device_type, verbose)
+
+
+if __name__ == "__main__":
+    device_select(platfrom_id=None, device_id=None, verbose=True)
```

Comment:
Is this necessary? I would have thought this was just for your testing and never actually gets called when SartOpencl() is used as the main entry point.

Reply:
This code is removed.